include/dlaf/factorization/qr/t_factor_impl.h

```diff
@@ -0,0 +1,347 @@
+
+            const SizeType first_element_in_tile = is_v0 ? x0.row() : 0;
+
+            // T(0:j, j) = -tau . V(j:, 0:j)* . V(j:, j)
+            // [j x 1] = [(n-j) x j]* . [(n-j) x 1]
+            TileElementSize va_size{tile_v.size().rows() - first_element_in_tile, x0.col()};
+            TileElementIndex va_start{first_element_in_tile, 0};
+            TileElementIndex vb_start{first_element_in_tile, x0.col()};
+
+            if (is_v0) {
+              tile_t(x0) = tau;
+
+              // use implicit 1 for the 2nd operand
+              for (SizeType r = 0; !va_size.isEmpty() && r < va_size.cols(); ++r) {
+                const TileElementIndex i_v{va_start.row(), r + va_start.col()};
+                const TileElementIndex i_t{r + t_start.row(), t_start.col()};
+
+                tile_t(i_t) = -tau * dlaf::conj(tile_v(i_v));
+              }
+
+              // skip already managed computations with implicit 1
+              va_start = va_start + TileElementSize{1, 0};
+              vb_start = vb_start + TileElementSize{1, 0};
+              va_size = {va_size.rows() - 1, va_size.cols()};
+            }
+
+            if (!va_size.isEmpty()) {
+              blas::gemv(blas::Layout::ColMajor, blas::Op::ConjTrans, va_size.rows(), va_size.cols(),
+                         -tau, tile_v.ptr(va_start), tile_v.ld(), tile_v.ptr(vb_start), 1, 1,
+                         tile_t.ptr(t_start), 1);
+            }
+          }
+          return std::move(tile_t);
+        });
+    return pika::dataflow(getHpExecutor<Backend::MC>(), gemv_func, tile_vi, taus, tile_t);
+  }
+
+  static pika::future<matrix::Tile<T, Device::CPU>> trmvUpdateColumn(
+      pika::future<matrix::Tile<T, Device::CPU>>& tile_t) noexcept {
+    // Update each column (in order) t = T . t
+    // remember that T is upper triangular, so it is possible to use TRMV
+    auto trmv_func = pika::unwrapping([](auto&& tile_t) {
+      for (SizeType j = 0; j < tile_t.size().cols(); ++j) {
+        const TileElementIndex t_start{0, j};
+        const TileElementSize t_size{j, 1};
+
+        blas::trmv(blas::Layout::ColMajor, blas::Uplo::Upper, blas::Op::NoTrans, blas::Diag::NonUnit,
+                   t_size.rows(), tile_t.ptr(), tile_t.ld(), tile_t.ptr(t_start), 1);
+      }
+      return std::move(tile_t);
+    });
+    return pika::dataflow(getHpExecutor<Backend::MC>(), trmv_func, tile_t);
+  }
+};
+
+#ifdef DLAF_WITH_CUDA
+template <class T>
+struct Helpers<Backend::GPU, Device::GPU, T> {
+  static pika::future<matrix::Tile<T, Device::GPU>> set0(pika::future<matrix::Tile<T, Device::GPU>>& t) {
+    return pika::dataflow(dlaf::cuda::Executor{dlaf::internal::getHpCudaStreamPool()},
+                          pika::unwrapping([](auto&& tile, cudaStream_t stream) {
+                            tile::internal::set0<T>(tile, stream);
+                            return std::move(tile);
+                          }),
+                          std::move(t));
+  }
```

Comment:
This is duplicated only because of the cublas/cusolver executors not working with plain streams, correct? Moving to the sender adaptors should allow you to use only one implementation.

Reply:
Not really. The executor is one reason, on the other hand as I need to define a task, it needs to be duplicated as the lambda has to have a different signatures:
CPU: [](tile)
GPU: [](tile, stream)

Reply:
You can define a helper
```
struct set0_but_also_return_tile {
    template <typename T, Device device, typename... Ts>
    auto operator()(Tile<T, device>&& tile, Ts&&... ts) {
        tile::internal::set0<T>(tile, std::forward<Ts>(ts)...);
        return std::move(tile);
    }
};
```
This is more or less what the `MAKE_CALLABLE_OBJECT` stuff does. However, only with the sender adaptors...

The second case below is clearly more tricky because it uses both plain cuda streams and cublas handles :/

Reply:
It might also be possible to create some sort of adaptor that would make this easier, a bit like `unwrapExtendTiles`. The downside of `unwrapExtendTiles` is that it returns a tuple when often that's unnecessary, or you don't care about the return value of the function that you're wrapping. I'll try to think about this a bit...

Reply:
I considered the variadic template solution, but IMO for these simple use cases I preferred to duplicate it.

Moreover, this solution works with the "stream at the end" case, but not with the "handle at the beginning" case.

Reply:
> It might also be possible to create some sort of adaptor that would make this easier, a bit like `unwrapExtendTiles`. The downside of `unwrapExtendTiles` is that it returns a tuple when often that's unnecessary, or you don't care about the return value of the function that you're wrapping. I'll try to think about this a bit...

One solution can be to return in all the cases the modified tile (e.g. gemm returns c, copy returns the destination tile, ...) but I'm not sure it is the correct approach.

Reply:
> I considered the variadic template solution, but IMO for these simple use cases I preferred to duplicate it.
> 
> Moreover, this solution works with the "stream at the end" case, but not with the "handle at the beginning" case.

Fair enough!

> One solution can be to return in all the cases the modified tile (e.g. gemm returns c, copy returns the destination tile, ...) but I'm not sure it is the correct approach.

Indeed, I'm also not sure it's the right thing to do. It somewhat goes against the principle of (when possible) accessing tiles through a matrix/panel, but then again it's clearly useful in some situations... Anyway, I'll close this as resolved as a refactoring would anyway require more work elsewhere first.